### PR TITLE
the result of decision procedure is fixed once witnessed is provided

### PR DESCRIPTION
**Pull request by HuStmpHrrr**
c.f. #781

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -644,3 +644,17 @@ Other minor additions
                           (isIdempotentCommutativeMonoid : IsIdempotentCommutativeMonoid ∙ ε)
 
   ```
+
+* Added a new module `Relation.Nullary.Irrelvant`:
+  ```agda
+  Irrelevant : ∀ {p} → Set p → Set p
+  Irrelevant P = ∀ (p₁ p₂ : P) → p₁ ≡ p₂
+  ```
```
Comment (gallais) [inline CHANGELOG.md]: There is a typo in `Irrelevant` in the module name.
Plus you can drop the type annotation in the CHANGELOG.

File: CHANGELOG.md [hunk 2]
```diff
@@ -644,3 +644,17 @@ Other minor additions
...
+
+* Added three lemmas in `Relation.Nullary.Decidable` which constraints the output of
+  decision procedures:
+  ```agda
+  p?-yes : ∀ (p? : Dec P) → P → ∃ λ p′ → p? ≡ yes p′
+  p?-no : ∀ (p? : Dec P) → ¬ P → ∃ λ ¬p′ → p? ≡ no ¬p′
+  p?-yes-irr : ∀ (p? : Dec P) → Irrelevant P → (p : P) → p? ≡ yes p
+  ```
```
Comment (gallais) [inline CHANGELOG.md]: Can you align the colons and drop the `∀`s please?

File: src/Relation/Nullary/Decidable.agda [hunk 3]
```diff
@@ -115,3 +116,20 @@ module _ {p} {P : Set p} where
   from-no : (p : Dec P) → From-no p
   from-no (no ¬p) = ¬p
   from-no (yes _) = _
+
+------------------------------------------------------------------------
+-- Result of decidability
+
+module _ {P : Set p} where
+
```
Comment (gallais) [inline src/Relation/Nullary/Decidable.agda]: Can you not drop the anonymous module? `P` is part of the variable block already!

File: src/Relation/Nullary/Decidable.agda [hunk 4]
```diff
@@ -115,3 +116,20 @@ module _ {p} {P : Set p} where
...
+  p?-yes (yes p′) p = p′ , refl
+  p?-yes (no ¬p) p = ⊥-elim (¬p p)
+
+  p?-no : ∀ (p? : Dec P) → ¬ P → ∃ λ ¬p′ → p? ≡ no ¬p′
+  p?-no (yes p) ¬p  = ¬p , ⊥-elim (¬p p)
+  p?-no (no ¬p′) ¬p = ¬p′ , refl
+
+  p?-yes-irr : ∀ (p? : Dec P) → Irrelevant P → (p : P) → p? ≡ yes p
+  p?-yes-irr p? irr p with p?-yes p? p
+  ... | p′ , eq rewrite irr p p′ = eq
```
Comment (gallais) [inline src/Relation/Nullary/Decidable.agda]: I am not sure about these names. I think I would use `dec` rather than `p?`.

Comment (gallais) [reply src/Relation/Nullary/Decidable.agda]: (Please also drop the `∀`s)

File: src/Relation/Nullary/Irrelevant.agda [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+------------------------------------------------------------------------
+
+{-# OPTIONS --without-K --safe #-}
+
+module Relation.Nullary.Irrelevant where
+
+open import Relation.Binary.PropositionalEquality
+
+Irrelevant : ∀ {p} → Set p → Set p
+Irrelevant P = ∀ (p₁ p₂ : P) → p₁ ≡ p₂
```
Comment (gallais) [inline src/Relation/Nullary/Irrelevant.agda]: I'm not sure this should be in its own module.
This is very reminiscent of the type of definitions we have in [`Relation.Binary.Core`](https://agda.github.io/agda-stdlib/Relation.Binary.Core.html)

Comment (MatthewDaggitt) [reply src/Relation/Nullary/Irrelevant.agda]: Yes, this should go in `Relation.Nullary`.

Comment (HuStmpHrrr) [reply src/Relation/Nullary/Irrelevant.agda]: no I can't. otherwise there is a cyclic dependency between PropositionalEquality and Nullary.

Comment (gallais) [reply src/Relation/Nullary/Irrelevant.agda]: You could import `Agda.Builtin.Equality` directly if you're only after the `_≡_` symbol.

**Comment by gallais** [comment]
It's worth refactoring `≡-≟-identity` and `≢-≟-identity` defined in
`Relation.Binary.PropositionalEquality` using these results.

**Comment by HuStmpHrrr** [comment]
> It's worth refactoring `≡-≟-identity` and `≢-≟-identity` defined in
> `Relation.Binary.PropositionalEquality` using these results.

this is hard. `Irrelevant` is defined by propositional equality and propositional equality depends on `Nullary`.

**Comment by HuStmpHrrr** [comment]
I can make `Irrelevant` into `Relation.Nullary` but importing `Relation.Nullary.Decidable` from `Relation.Binary.PropositionalEquality` is a bit hard to settle. these two modules are imported everywhere.

the cyclic imports are:

```
  Relation.Binary.PropositionalEquality
  Relation.Nullary.Decidable
  Function.Equivalence
  Relation.Binary.PropositionalEquality
```

**Comment by gallais** [comment]
Note that #608 introduces `Relation.Nullary.Decidable.Core` :D

**Comment by HuStmpHrrr** [comment]
i see, let's wait for your PR to get merged first.

**Comment by MatthewDaggitt** [comment]
Unblocked.

**Pull request merged**